gui/test/ruler-tests.cc

```diff
@@ -0,0 +1,370 @@
+{
+    BOOST_CHECK(RulerCalculations::intervalPixelSpacing(5, 0, 1000, 540) == 3);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_intervalPixelSpacing_range_0_to_1000_interval_5_size_1920px,
+     * utf::description("Tests that an interval of 5 for range [0, 1000] for a ruler of 1920px gives a spacing of 2px"))
+{
+    BOOST_CHECK(RulerCalculations::intervalPixelSpacing(5, 0, 1000, 1920) == 10);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_intervalPixelSpacing_range_0_to_1000_interval_25_size_540px,
+     * utf::description("Tests that an interval of 25 for range [0, 1000] for a ruler of 1920px gives a spacing of 1px"))
+{
+    BOOST_CHECK(RulerCalculations::intervalPixelSpacing(25, 0, 1000, 540) == 14);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_intervalPixelSpacing_range_0_to_1000_interval_25_size_1920px,
+     * utf::description("Tests that an interval of 25 for range [0, 1000] for a ruler of 1920px gives a spacing of 2px"))
+{
+    BOOST_CHECK(RulerCalculations::intervalPixelSpacing(25, 0, 1000, 1920) == 48);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_intervalPixelSpacing_range_0_to_1000_interval_50000000_size_540px,
+     * utf::description("Tests that an interval of 25 for range [0, 1000] for a ruler of 1920px gives a spacing of 1px"))
+{
+    BOOST_CHECK(RulerCalculations::intervalPixelSpacing(50000000, 0, 1000000000, 540) == 27);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_intervalPixelSpacing_range_0_to_1000_interval_50000000_size_1920px,
+     * utf::description("Tests that an interval of 25 for range [0, 1000] for a ruler of 1920px gives a spacing of 2px"))
+{
+    BOOST_CHECK(RulerCalculations::intervalPixelSpacing(50000000, 0, 1000000000, 1920) == 96);
+}
+
+
+///////////////
+// Testing firstTick()
+
+
+
+BOOST_AUTO_TEST_CASE(Ruler_firstTick_lowerLimit_0_interval_1,
+     * utf::description("Tests that a lower limit of 0 and interval 1 gives first tick position of 0"))
+{
+    BOOST_CHECK(RulerCalculations::firstTick(0, 1) == 0);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_firstTick_lowerLimit_0_interval_25,
+     * utf::description("Tests that a lower limit of 0 and interval 25 gives first tick position of 0"))
+{
+    BOOST_CHECK(RulerCalculations::firstTick(0, 25) == 0);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_firstTick_lowerLimit_0_interval_50000,
+     * utf::description("Tests that a lower limit of 0 and interval 50000 gives first tick position of 0"))
+{
+    BOOST_CHECK(RulerCalculations::firstTick(0, 50000) == 0);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_firstTick_lowerLimit_neg123_interval_1,
+     * utf::description("Tests that a lower limit of -123 and interval 1 gives first tick position of -123"))
+{
+    BOOST_CHECK(RulerCalculations::firstTick(-123, 1) == -123);
+}
+
+BOOST_AUTO_TEST_CASE(Ruler_firstTick_lowerLimit_neg123_interval_25,
+     * utf::description("Tests that a lower limit of -123 and interval 25 gives first tick position of -125"))
+{
+    BOOST_CHECK(RulerCalculations::firstTick(-123, 25) == -125);
```

Comment:
first tick can be before the start of the range?

Reply:
Yes, it usually starts drawing from outside the drawing area, so it can still draw the minor ticks at the start, because the ruler doesn't always start with a major tick. This seemed like the simplest solution to me, instead of adding a separate case for drawing the minor ticks at the start.